README.md
# radixmap - Map à la Binary Radix Tree

This OCaml library provides a pure map-like data structure which efficiently
represent piecewise constant functions over domains which can be binary
subdivided.  Suitable domains include bit strings, byte strings, and IP
addresses.  It is most efficient where the maps tend to be piecewise
constant over prefixes of the key type, with optional nested exceptions
which themselves have the same structure.

This package has an optional dependency on
[ipaddr](https://github.com/mirage/ocaml-ipaddr) and provides a related set
implementation `Ip_radixset` if present.

The core map, `Bitword_radixmap`, works on finite size `Bitword` sub-paths,
whereas longer paths are encoded in the tree itself.  This makes the maps
compact for common cases.  A higher level map or set can be implemented in
terms of `Bitword_radixmap` if keys may exceed 24 bits.


CHANGES.md
# Changelog

There is no release yet.


tests/testkit.ml
(* Copyright (C) 2017  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

open Printf

let init test_name =
  let seed =
    try int_of_string (Unix.getenv "SEED")
    with Not_found -> int_of_float (Unix.time ()) in
  Random.init seed;
  printf "Running %s with seed %d (keep if you hit a bug).\n"
         test_name seed


tests/test_bitword.ml
(* Copyright (C) 2017  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

module Test_endian (Xe : sig include Bitword.ENDIAN_API val is_be : bool end) =
struct
  let test_one w =
    let n = Bitword.length w in
    assert (w = Xe.init n (Xe.get w));
    for i = 0 to n do
      let pfx = Xe.prefix i w in
      let sfx = Xe.suffix i w in
      let pfx', sfx' = Xe.cut i w in
      assert (Bitword.length pfx = (if Xe.is_be then i else n - i));
      assert (Bitword.length sfx = (if Xe.is_be then n - i else i));
      assert (pfx = pfx');
      assert (sfx = sfx');
      assert (Bitword.cat_exn pfx sfx = w)
    done;
    let w' = Xe.drop_exn w in
    assert (Xe.drop_unsafe w = w');
    let b, v = Xe.pop_exn w in
    let b', v' = Xe.pop_unsafe w in
    assert (b = b');
    assert (v = v');
    assert (v = w');
    assert (Xe.push_exn b v = w);
    let push_cb_exn, push_cb_unsafe =
      if b then Xe.(push_c1_exn, push_c1_unsafe)
           else Xe.(push_c0_exn, push_c0_unsafe) in
    assert (push_cb_exn v = w);
    assert (push_cb_unsafe v = w)

  let test () =
    for n = 1 to Bitword.max_length do
      for round = 1 to Pervasives.(min 10000 (1 lsl n)) do
        test_one (Bitword.random_uniform n)
      done
    done
end

module Test_be = Test_endian (struct include Bitword.Be let is_be = true end)
module Test_le = Test_endian (struct include Bitword.Le let is_be = false end)

let () =
  Testkit.init "test_bitword";

  assert (Bitword.max_length mod 8 = 0);
  assert (Bitword.max_length >= 24);
  assert (Bitword.length Bitword.c0 = 1);
  assert (Bitword.length Bitword.c1 = 1);
  assert (Bitword.Le.get Bitword.c0 0 = false);
  assert (Bitword.Le.get Bitword.c1 0 = true);
  assert (Bitword.bits Bitword.c0 = 0);
  assert (Bitword.bits Bitword.c1 = 1);
  assert (Bitword.equal Bitword.c0 Bitword.c0);
  assert (Bitword.equal Bitword.c1 Bitword.c1);
  assert (not (Bitword.equal Bitword.c0 Bitword.c1));
  assert (not (Bitword.equal Bitword.c1 Bitword.c0));
  assert (Bitword.compare Bitword.c0 Bitword.c0 = 0);
  assert (Bitword.compare Bitword.c1 Bitword.c1 = 0);
  assert (Bitword.compare Bitword.c0 Bitword.c1 = -1);
  assert (Bitword.compare Bitword.c1 Bitword.c0 = 1);
  assert (Bitword.empty <> Bitword.c0);
  assert (Bitword.empty <> Bitword.c1);

  assert (Bitword.const 0 false = Bitword.empty);
  assert (Bitword.const 0 true  = Bitword.empty);
  assert (Bitword.const 1 false = Bitword.c0);
  assert (Bitword.const 1 true  = Bitword.c1);
  assert (Bitword.const_unsafe 0 false = Bitword.empty);
  assert (Bitword.const_unsafe 0 true  = Bitword.empty);
  assert (Bitword.const_unsafe 1 false = Bitword.c0);
  assert (Bitword.const_unsafe 1 true  = Bitword.c1);

  for n = 0 to Bitword.max_length do
    let w0 = Bitword.const n false in
    let w1 = Bitword.const n true in
    assert (Bitword.length w0 = n);
    assert (Bitword.length w1 = n);
    assert (w0 = Bitword.Le.init n (fun _ -> false));
    assert (w1 = Bitword.Le.init n (fun _ -> true));
    assert (w0 = Bitword.make n 0);
    assert (w1 = Bitword.make n Pervasives.(1 lsl n - 1));
    for round = 0 to Pervasives.(min 10000 (1 lsl n)) do
      let w = Bitword.random_uniform n in
      assert (Bitword.length w = n);
      assert (w = Bitword.make n (Bitword.bits w));

      (* common prefix *)
      let v = Bitword.random_uniform (Random.int (n + 1)) in
      let l = Bitword.common_prefix_length w v in
      assert (Bitword.Be.prefix l w = Bitword.Be.prefix l v);
      assert (l = Bitword.length v || Bitword.Be.get v l <> Bitword.Be.get w l);

      (* cat_rem *)
      let v = Bitword.random_uniform (Random.int (Bitword.max_length + 1)) in
      let v', w' = Bitword.cat_rem v w in
      assert Bitword.(length v' = 0 || length w' = max_length);
      assert Bitword.(length v' + length w' = length v + length w);
      let nw = Bitword.length w in
      let nv = Bitword.length v in
      for i = 0 to nw - 1 do
        assert (Bitword.Le.get w i = Bitword.Le.get w' i)
      done;
      let n' = min nv (Bitword.max_length - nw) in
      for i = 0 to n' - 1 do
        assert (Bitword.Le.get v i = Bitword.Le.get w' (nw + i))
      done;
      for i = n' to nv - 1 do
        assert (Bitword.Le.get v i = Bitword.Le.get v' (i - n'))
      done
    done
  done;

  Test_be.test ();
  Test_le.test ()


myocamlbuild.ml
(* Copyright (C) 2017  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

open Ocamlbuild_plugin

let () =
  rule "%.mli & %.idem -> %.ml"
    ~deps:["%.mli"; "%.idem"] ~prod:"%.ml"
    (fun env _ -> cp (env "%.mli") (env "%.ml"))


lib/ip_radixset.ml
(* Copyright (C) 2017  Petter A. Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

module type BASE = sig
  type t
  type address
  type network

  val max_length : int
  val bytes_of_address : address -> string
  val bytes_of_network : network -> string
  val length_of_network : network -> int

  val is_full : t -> bool
  val of_network : network -> t
  val contains_address : t -> address -> bool
  val contains_network : t -> network -> bool
  val make_index : int -> Bytes.t -> network
end

module Bool_equal = struct
  type t = bool
  let equal (x : t) (y : t) = x = y
end

module M = Bitword_radixmap.Make (Bool_equal)
open M

module V4_base = struct
  type t = M.t
  type address = Ipaddr.V4.t
  type network = Ipaddr.V4.Prefix.t

  let max_length = 32
  let bytes_of_address addr = Ipaddr.V4.to_bytes addr
  let bytes_of_network net = Ipaddr.V4.to_bytes (Ipaddr.V4.Prefix.network net)
  let length_of_network net = Ipaddr.V4.Prefix.bits net

  let is_full s = value s = Some true

  let of_network pfx =
    let l = Ipaddr.V4.Prefix.bits pfx in
    let netH, netL = Ipaddr.V4.Prefix.network pfx |> Ipaddr.V4.to_int16 in
    if l <= 16 then
      const true |> unzoom false (Bitword.make l (netH lsr (16 - l)))
    else
      const true |> unzoom false (Bitword.make (l - 16) (netL lsr (32 - l)))
                 |> unzoom false (Bitword.make 16 netH)

  let contains_address s addr =
    let aH, aL = Ipaddr.V4.to_int16 addr in
    s |> zoom (Bitword.make 16 aH)
      |> zoom (Bitword.make 16 aL)
      |> is_full

  let contains_network s pfx =
    let l = Ipaddr.V4.Prefix.bits pfx in
    let aH, aL = Ipaddr.V4.Prefix.network pfx |> Ipaddr.V4.to_int16 in
    if l <= 16 then
      s |> zoom (Bitword.make l (aH lsr (16 - l)))
        |> is_full
    else
      s |> zoom (Bitword.make 16 aH)
        |> zoom (Bitword.make (l - 16) (aL lsr (32 - l)))
        |> is_full

  let make_index plen pbuf =
    Ipaddr.V4.Prefix.make plen (Ipaddr.V4.of_bytes_exn pbuf)
end

module V6_base = struct
  type t = M.t
  type address = Ipaddr.V6.t
  type network = Ipaddr.V6.Prefix.t

  let max_length = 128
  let bytes_of_address addr = Ipaddr.V6.to_bytes addr
  let bytes_of_network net = Ipaddr.V6.to_bytes (Ipaddr.V6.Prefix.network net)
  let length_of_network net = Ipaddr.V6.Prefix.bits net

  let is_full s = value s = Some true

  let of_network pfx =
    let l = Ipaddr.V6.Prefix.bits pfx in
    let net = Ipaddr.V6.Prefix.network pfx |> Ipaddr.V6.to_bytes in
    let rec loop i acc =
      if i < 0 then acc else
      loop (i - 1) (acc |> unzoom false (Bitword.make 8 (Char.code net.[i]))) in
    let lr = l mod 8 in
    if lr = 0 then
      loop (l / 8 - 1) (const true)
    else
      let w = Bitword.make lr (Char.code net.[l / 8] lsr (8 - lr)) in
      loop (l / 8 - 1) (const true |> unzoom false w)

  let contains_address s addr =
    let a0, a1, a2, a3, a4, a5, a6, a7 = Ipaddr.V6.to_int16 addr in
    s |> zoom (Bitword.make 16 a7) |> zoom (Bitword.make 16 a6)
      |> zoom (Bitword.make 16 a5) |> zoom (Bitword.make 16 a4)
      |> zoom (Bitword.make 16 a3) |> zoom (Bitword.make 16 a2)
      |> zoom (Bitword.make 16 a1) |> zoom (Bitword.make 16 a0)
      |> is_full

  let contains_network s pfx =
    let l = Ipaddr.V6.Prefix.bits pfx in
    let a = Ipaddr.V6.to_bytes (Ipaddr.V6.Prefix.network pfx) in
    let rec loop k acc =
      if k = l then acc else
      if k + 8 > l then
        let lr = l mod 8 in
        acc |> zoom (Bitword.make lr (Char.code a.[k / 8] lsr (8 - lr)))
      else
        loop (k + 8) (acc |> zoom (Bitword.make 8 (Char.code a.[k / 8]))) in
    loop 0 s |> is_full

  let make_index plen pbuf =
    Ipaddr.V6.Prefix.make plen (Ipaddr.V6.of_bytes_exn pbuf)
end

module Make (Base : BASE) = struct
  include Base

  let empty = const false
  let is_empty s = value s = Some false
  let full = const true

  let equal = M.equal

  let flip_address addr x s =
    let b = bytes_of_address addr in
    let rec loop i =
      if i = max_length / 8 then (fun _ -> const x) else
      let w = Bitword.make 8 (Char.code b.[i]) in
      modify w (loop (i + 1)) in
    loop 0 s

  let flip_network net x s =
    let l = length_of_network net in
    let b = bytes_of_network net in
    let rec loop k =
      if k = l then (fun _ -> const x) else
      if k + 8 > l then
        let w = Bitword.make (l - k) (Char.code b.[k / 8] lsr (8 - l + k)) in
        modify w (fun _ -> const x)
      else
        let w = Bitword.make 8 (Char.code b.[k / 8]) in
        modify w (loop (k + 8)) in
    loop 0 s

  let add_address addr s = flip_address addr true s
  let remove_address addr s = flip_address addr false s
  let add_network net s = flip_network net true s
  let remove_network net s = flip_network net false s

  let union sA sB = merge (||) sA sB
  let inter sA sB = merge (&&) sA sB
  let compl sX sU = merge (fun x y -> not x && y) sX sU

  let catai ~const ~appose ~unzoom s =
    catai_bytes
      ~index_buffer_size:(max_length / 8)
      ~make_index ~const ~appose ~unzoom s

  let rec is_network s =
    recurse
      ~const:(fun _ -> false)
      ~appose:(fun _ _ -> false)
      ~unzoom:(fun x p s' -> not x && is_network s') s

  let valid = M.valid
end

module V4 = Make (V4_base)
module V6 = Make (V6_base)


lib/bitword.ml
(* Copyright (C) 2017  Petter Urkedal <paurkedal@gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version, with the OCaml static compilation exception.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *)

type t = int

(* Representation *)

(* TODO: Need a compile-time constant for efficient switch to 56 bit width for
 * 64 bit native compilation. *)
let length_width = 5
let max_length = 24
let floor_log2_halfwidth = 16

let bit0 = 1 lsl length_width
let length_mask = bit0 - 1
let bitmask = (1 lsl max_length - 1) lsl length_width

(* Helpers *)

(* Lifted from prime_int.ml. *)
let rec floor_log2_loop j n l =
  if j = 0 then l else
  if n lsr j = 0
    then floor_log2_loop (j / 2) n l
    else floor_log2_loop (j / 2) (n lsr j) (l + j)
let floor_log2 n = floor_log2_loop floor_log2_halfwidth n 0

(* Inspection and Slicing *)

let length p = p land length_mask

let bits p = p lsr length_width

(* Comparison *)

let equal (v : int) (w : int) = v = w

let compare p q =
  let m, n = length p, length q in
  let p' = p lsl (max_length - m) land bitmask in
  let q' = q lsl (max_length - n) land bitmask in
  if p' < q' then -1 else
  if p' > q' then  1 else
  if m  < n then -1 else if n > m then 1 else 0

let common_prefix_length p q =
  let m, n = length p, length q in
  if m < n
  then length_width - 1 + m - floor_log2 (p lxor q lsr (n - m) lor length_mask)
  else length_width - 1 + n - floor_log2 (q lxor p lsr (m - n) lor length_mask)

(* Construction *)

let is_empty p = p = 0
let is_full p = length p = max_length

let empty = 0
let c0 = 1
let c1 = 1 lor bit0
let c00 = 2
let c01 = 2 lor bit0
let c10 = 2 lor 2 lsl length_width
let c11 = 2 lor 3 lsl length_width

let make_unsafe n x = x lsl length_width lor n
let make n x =
  if n < 0 || n > max_length || x lsr n <> 0 then invalid_arg "Bitword.make";
  x lsl length_width lor n

let const_unsafe n x = (if x then (1 lsl n - 1) lsl length_width else 0) lor n
let const n x =
  if n < 0 || n > max_length then invalid_arg "Bitword.const" else
  const_unsafe n x

let random_uniform n =
  if n < 0 || n > max_length then invalid_arg "Bitword.random_uniform";
  let rec loop m acc =
    if m <= 0 then acc else
    loop (m - 30) (acc lsl 30 lor Random.bits ()) in
  (loop n 0 lsl length_width) land (1 lsl n - 1) lor n

(* Endian-Dependent Operations *)

module type ENDIAN_API = sig
  val init : int -> (int -> bool) -> t
  val get : t -> int -> bool

  val prefix : int -> t -> t
  val prefix_unsafe : int -> t -> t
  val suffix : int -> t -> t
  val suffix_unsafe : int -> t -> t

  val cut : int -> t -> t * t

  val push_exn : bool -> t -> t
  val push_c0_exn : t -> t
  val push_c0_unsafe : t -> t
  val push_c1_exn : t -> t
  val push_c1_unsafe : t -> t
  val drop_exn : t -> t
  val drop_unsafe : t -> t
  val pop_exn : t -> bool * t
  val pop_unsafe : t -> bool * t
end

module Be = struct

  let init n f =
    let rec loop i acc =
      if i = n then acc lsl length_width lor n else
      loop (i + 1) (acc lsl 1 lor (if f i then 1 else 0)) in
    if n < 0 || n > max_length then invalid_arg "Bitword.Be.init" else
    loop 0 0

  let get p i = p lsr (length p - 1 - i) land bit0 <> 0

  let prefix_unsafe k p =
    let n = length p in
    (p lsr (n - k) land bitmask) lor k

  let prefix k p =
    let n = length p in
    if k < 0 || k > n then invalid_arg "Bitword.Be.prefix" else
    (p lsr (n - k) land bitmask) lor k

  let suffix_unsafe k p =
    let n = length p in
    p land (1 lsl (n - k) - 1) lsl length_width lor (n - k)

  let suffix k p =
    let n = length p in
    if k < 0 || k > n then invalid_arg "Bitword.Be.suffix" else
    p land (1 lsl (n - k) - 1) lsl length_width lor (n - k)

  let cut k p =
    let n = length p in
    if k < 0 || k > n then invalid_arg "Bitword.Be.cut" else
    (prefix_unsafe k p, suffix_unsafe k p)

  let push_exn x p =
    let n = length p in
    if n >= max_length then invalid_arg "Bitword.Be.push_exn" else
    p lor (if x then bit0 lsl n else 0) + 1

  let push_c0_unsafe p = p + 1
  let push_c0_exn p =
    if length p >= max_length then invalid_arg "Bitword.Be.push_c0_exn" else
    push_c0_unsafe p

  let push_c1_unsafe p = p lor bit0 lsl length p + 1
  let push_c1_exn p =
    if length p >= max_length then invalid_arg "Bitword.Be.push_c1_exn" else
    push_c1_unsafe p

  let drop_unsafe p =
    let n = length p in
    p land (bit0 lsl (n - 1) - 1) - 1

  let drop_exn p =
    let n = length p in
    if n = 0 then invalid_arg "Bitword.Be.drop_exn" else
    p land (bit0 lsl (n - 1) - 1) - 1

  let pop_unsafe p =
    let n = length p in
    let bit = bit0 lsl (n - 1) in
    (p land bit <> 0, p land (lnot bit) - 1)

  let pop_exn p =
    let n = length p in
    if n = 0 then invalid_arg "Bitword.Be.pop_exn" else
    let bit = bit0 lsl (n - 1) in
    (p land bit <> 0, p land (lnot bit) - 1)
end

module Le = struct

  let init n f =
    let rec loop i acc =
      if i = n then acc lsl length_width lor n else
      loop (i + 1) (if f i then acc lor 1 lsl i else acc) in
    if n < 0 || n > max_length then invalid_arg "Bitword.Le.init" else
    loop 0 0

  let get p i = p lsr i land bit0 <> 0

  let prefix_unsafe k p =
    let n = length p in
    p lsr k land bitmask lor (n - k)

  let prefix k p =
    let n = length p in
    if k < 0 || k > n then invalid_arg "Bitword.Le.prefix" else
    p lsr k land bitmask lor (n - k)

  let suffix_unsafe k p =
    p land (1 lsl k - 1) lsl length_width lor k

  let suffix k p =
    let n = length p in
    if k < 0 || k > n then invalid_arg "Bitword.Le.suffix" else
    suffix_unsafe k p

  let cut k p =
    let n = length p in
    if k < 0 || k > n then invalid_arg "Bitword.Le.cut" else
    (p lsr k land bitmask lor (n - k), suffix_unsafe k p)

  let push_exn x p =
    let n = length p in
    if n >= max_length then invalid_arg "Bitword.Le.push_exn" else
    (p land bitmask) lsl 1 lor (if x then bit0 else 0) lor (n + 1)

  let push_c0_unsafe p = (p land bitmask) lsl 1 lor (length p + 1)
  let push_c0_exn p =
    if length p >= max_length then invalid_arg "Bitword.Le.push_c0_exn" else
    push_c0_unsafe p

  let push_c1_unsafe p = (p land bitmask) lsl 1 lor bit0 lor (length p + 1)
  let push_c1_exn p =
    if length p >= max_length then invalid_arg "Bitword.Le.push_c1_exn" else
    push_c1_unsafe p

  let drop_unsafe p = p lsr 1 land bitmask lor (length p - 1)
  let drop_exn p =
    if length p = 0 then invalid_arg "Bitword.Le.drop_exn" else
    drop_unsafe p

  let pop_unsafe p = (p land bit0 <> 0, drop_unsafe p)
  let pop_exn p =
    if length p = 0 then invalid_arg "Bitword.Le.pop_exn" else
    pop_unsafe p
end

(* Slice and Concatenate *)

let cat_exn p q =
  let m = length p in
  let n = length q in
  if n + m > max_length then invalid_arg "Bitword.cat_exn" else
  (p land bitmask) lsl n lor q + m

let cat_rem p q =
  let m = length p in
  let n = length q in
  let k = min m (max_length - n) in
  let p' = (p lsr k land bitmask) + (m - k) in
  let q' = ((p land bitmask) lsl n land bitmask) lor q + k in
  (p', q')

(* Conversion and Pretty-Printing *)

let pp_set fmtr w =
  let rec loop i u =
    if u <> 0 then begin
      if u land 1 <> 0 then begin
        Format.pp_print_int fmtr i;
        Format.pp_print_string fmtr ", "
      end;
      loop (i + 1) (u lsr 1)
    end in
  Format.pp_print_char fmtr '{';
  loop 0 (w lsr length_width);
  Format.pp_print_char fmtr '}'

let to_base2_string p =
  let n = length p in
  String.init n (fun i -> if Le.get p (n - i - 1) then '1' else '0')

let pp_base2 fmtr p = Format.pp_print_string fmtr (to_base2_string p)

let of_base2_string_exn s =
  let n = String.length s in
  if n > max_length then invalid_arg "Bitword.of_base2_string_exn";
  Le.init n
    (fun i ->
      (match s.[i] with
       | '0' -> false
       | '1' -> true
       | _ -> invalid_arg "Bitword.of_base2_string_exn"))
